### ISSUE TITLE
What's the update strategy for streaming jobs?

### ISSUE THREAD

**Issue by casidiablo** [spotify/scio#163]
When using plain Java, and if you religiously name every step in the job, upgrading a pipeline can be easy as explained in the official documentation:

https://cloud.google.com/dataflow/pipelines/updating-a-pipeline

Using scio, however, there does not seem to be a way to name steps, especially those created by methods in the SCollection. This is rather inconvenient because the job scheduler names the steps with things like `apply@{SCollection.scala:335}` or `groupBy@{MyClass.scala:93}2`; which means that if you update the scio library or, even worse, slightly change your own code, the new pipeline will no longer be compatible.

I've even tried to name the steps using `.applyTransform(ParDo.named("SaveToS3")...)`, but the resulting step still has the same issue (in this case it was named `applyTransform@{MyClass.scala:140}`).

What is the updating strategy used at Spotify to address this kind of scenarios? Creating a new pipeline is, of course, not an option, since data could be loss.

**Comment by nevillelyh** [comment]
Great question! Most internal jobs are batch with a few guys experimenting with streaming so updating streaming job is not a big concern right now.

There are several options I can think of:
- Some flag to disable transform auto naming e.g. `map@{Job.scala:123}` but we still need to name them consistently
- Overload every method with `(name: String, args...)` but that's messy
- Implement a `@name` annotation so one can do `coll.map(@name("Input") fn)`. I'm not an annotation expert though so will have to play around with this idea first.

What do you think?

**Comment by casidiablo** [comment]
The annotations seem to be the best way to go, definitely; not sure how easy it'd be to implement, though.

I think, unfortunately, I'm going to have to go back to a pure Java job in the meantime. I wish I had thought about this before implementing the whole thing in Scala. I will keep my eyes on this issue to see how this evolves.

Thank you guys for the hard work.

**Comment by nevillelyh** [comment]
Looks like one can only annotate methods or arguments when defining them and not on invocation so the 3rd options is out.

**Comment by nevillelyh** [comment]
Igor mentioned that there's an experimental drain feature that might solve this problem.

**Comment by nevillelyh** [comment]
https://cloud.google.com/blog/big-data/2016/08/managing-streaming-pipelines-in-google-cloud-dataflow-just-got-better

@casidiablo Google just announced this "drain" feature. Not exactly the same as update but should solve the problem to some extent?

**Comment by ElPicador** [comment]
Having the update feature in scio will really be a big improvement for us.
We will mostly have streaming jobs, with 1 day windows. So having an update in place is really important for us.

I kind of like the `Overload every method with (name: String, args...)` it's simple

**Comment by ElPicador** [comment]
Could we help you on this one? We are willing to code it. Have you any idea on who you would want to tackle this issue?

Because the "drain" feature doesn't suit our need.

**Comment by nevillelyh** [comment]
@ElPicador there are several solutions that you could try.

- First option is to overload every method and add `name: String` argument, preferably as the 1st arg since otherwise it'll break methods that have args with default values.

We didn't do this because it's a massive change, and does not guarantee stable names for every transform. There are still some used inside high level ones like `join`, `cross`, etc.

- Change `CallSites.getCurrent` so that it doesn't include filename & line no, e.g. from `map@{Job.scala:123}` to `map#n` where `n` is a unique ID that's generated more deterministically. This won't help for cases like replacing `map` with `flatMap` though.

Personally I prefer this option if we can come up with a more predictable way of automatic naming.

- I looked into annotations but don't think it solves our problem.

What do you think?

**Comment by ElPicador** [comment]
I like the idea to change `CallSites.getCurrent`, as it seems a simple change.

I'll try to come up with something, and make a PR

**Comment by ElPicador** [comment]
After talking (again) within the team, a good night sleep, and careful (rere) reading of [dataflow doc](https://cloud.google.com/dataflow/pipelines/updating-a-pipeline), I think this deterministic but automatic naming won't resolve the issue of updating jobs.

The dataflow doc is pretty clear on this, steps must be named to update them. So it's mandatory to have user specific names for them.

My best (but not ideal) idea right now, is to add a `name` method to `SCollection` that takes a `String` and name the next step with it: `scollection.name("my step").map(...)`
I agree it's bad but I don't see a better way without breaking scio API.

Updating streaming jobs is a must have feature for us, we :hear: scio, but it could be a reason for us to drop it :(

As I said before we are willing to help, specially on this feature.

**Comment by nevillelyh** [comment]
So can we agree on that deterministic naming will solve the case of changing bevaior inside an `fn`, as long as the transform names are stable, e.g. `pubsubTopic`, `map`, `filter`, `saveAsBigQueryTable`?

However it won't help changing transform types, e.g. `pubsubTopic` &rarr; `kafkaTopic`, `map` &rarr; `flatMap`. Explicitly naming them might solve this but not the case where you replace `map`+`filter` with `flatMap`, essentially a graph change. Idiomatic Scala is by design a lot more granular than `apply()` in the Java SDK.

And there are other more complex graph changes that the native Java API can't solve either, like adding side input, replacing join with side input, right?

However have you looked at the other angle, using [transformNameMapping](https://cloud.google.com/dataflow/pipelines/updating-a-pipeline#Mapping)? This is a parameter when submitting jobs so we can compute it at runtime. We can do something like this:
- Add a dry-run option to export step names
- Dry-run old and new pipeline
- Diff list of steps to compute mapping. This should work if there isn't drastic code change. Worse case can be done manually, but just as manual as explicitly naming steps :)
- Submit new pipeline with mapping

Thoughts?

**Comment by ElPicador** [comment]
Using the dry-run-export-diff could work, it's a bit painful, and not really related to scio.
It seems there is no easy way to add this functionality without breaking scio API.
For now, in our project, we will fork the code of `SCollection` to support methods with a `name: String` as a first parameter. 

It's a bit more work on our side but I don't see any other options :/

**Comment by nevillelyh** [comment]
This just occurred to me. We can wrap most lambdas in a `named` variation with a `name: String` field to be retrieved inside later. For example:

```scala
trait NamedFunction extends Serializable {
  val name: String
}

object NamedFunctions {
  trait NamedFunction1[T1, R] extends (T1 => R) with NamedFunction
  trait NamedFunction2[T1, T2, R] extends ((T1, T2) => R) with NamedFunction
}

trait Named {
  import NamedFunctions._

  def named[T1, R](n: String)(f: T1 => R) = new NamedFunction1[T1, R] {
    private val g = f
    override val name: String = n
    override def apply(v1: T1): R = g(v1)
  }

  def named2[T1, T2, R](n: String)(f: (T1, T2) => R) = new NamedFunction2[T1, T2, R] {
    private val g = f
    override val name: String = n
    override def apply(v1: T1, v2: T2): R = g(v1, v2)
  }
}
```

The `com.spotify.scio` package object can extend `trait Named` so that `import com.spotify.scio._` to include all `named{n}` methods.

And pipeline code would look like this:

```scala
sc.textFile("in.txt")
  .flatMap(named("tokenize")(_.split(" ").toList))
```

There are some rough edges though. For example we'll need `name2`, `name3`, ... due to erasure. And in the `flatMap` case above `String => Array[String]` no longer works since it couldn't figure out `Array[T] => TraversableOnce[T]` implicit conversion.

But I still want to get your opinion first. @ElPicador @casidiablo thoughts?

**Comment by ElPicador** [comment]
Seems a pretty good idea. I'll try it

**Comment by ElPicador** [comment]
I've played a little with the idea, on the `MinimalWordCount` example

it goes from:
```
sc.textFile(args.getOrElse("input", ExampleData.KING_LEAR))
      .flatMap(_.split("[^a-zA-Z']+").filter(_.nonEmpty))
      .countByValue
      .map(t => t._1 + ": " + t._2)
      .saveAsTextFile(args("output"))
```

to

```
sc.textFile(args.getOrElse("input", ExampleData.KING_LEAR))
      .flatMap(namedFM("ExtractWords")(_.split("[^a-zA-Z']+").filter(_.nonEmpty)))
      .countByValue("Count words")
      .map(named("FormatResults")(t => t._1 + ": " + t._2))
      .saveAsTextFile(args("output"))
```

As you said there is a lot of edge cases. I'm not a big fan of it. It will add a lot of `namedXXXX` functions to make it work :/

WDYT?

**Comment by andrewsmartin** [comment]
@ElPicador you mentioned at first dropping down to using `applyTransform(ParDo.named(...))`, if we were to at least make that work (as it doesn't right now) would that be an OK compromise for now? It just means you will have to drop down to the Java SDK whenever you do want to specify a name. All other solutions we can think of would over-complicate the API.

**Comment by ElPicador** [comment]
It won't fit our need, but at least there is a way to do it

**Comment by andrewsmartin** [comment]
Closed via 535543e
